cmd/outline-ss-server/metrics.go

```diff
@@ -31,20 +31,138 @@ import (
+	}, nil
+}
+
+func (c *proxyCollector) Describe(ch chan<- *prometheus.Desc) {
+	c.dataBytesPerKey.Describe(ch)
+	c.dataBytesPerLocation.Describe(ch)
+}
+
+func (c *proxyCollector) Collect(ch chan<- prometheus.Metric) {
+	c.dataBytesPerKey.Collect(ch)
+	c.dataBytesPerLocation.Collect(ch)
+}
+
+func (c *proxyCollector) addOutbound(clientProxyBytes, proxyTargetBytes int64, accessKey string, clientInfo ipinfo.IPInfo) {
+	addIfNonZero(clientProxyBytes, c.dataBytesPerKey, "c>p", accessKey)
+	addIfNonZero(clientProxyBytes, c.dataBytesPerLocation, "c>p", clientInfo.CountryCode.String(), asnLabel(clientInfo.ASN))
+	addIfNonZero(proxyTargetBytes, c.dataBytesPerKey, "p>t", accessKey)
+	addIfNonZero(proxyTargetBytes, c.dataBytesPerLocation, "p>t", clientInfo.CountryCode.String(), asnLabel(clientInfo.ASN))
+}
+
+func (c *proxyCollector) addInbound(targetProxyBytes, proxyClientBytes int64, accessKey string, clientInfo ipinfo.IPInfo) {
+	addIfNonZero(targetProxyBytes, c.dataBytesPerKey, "p<t", accessKey)
+	addIfNonZero(targetProxyBytes, c.dataBytesPerLocation, "p<t", clientInfo.CountryCode.String(), asnLabel(clientInfo.ASN))
+	addIfNonZero(proxyClientBytes, c.dataBytesPerKey, "c<p", accessKey)
+	addIfNonZero(proxyClientBytes, c.dataBytesPerLocation, "c<p", clientInfo.CountryCode.String(), asnLabel(clientInfo.ASN))
+}
+
+type tcpConnMetrics struct {
+	tcpCollector        *tcpCollector
+	tunnelTimeCollector *tunnelTimeCollector
+
+	localAddr  net.Addr
+	clientAddr net.Addr
+	clientInfo ipinfo.IPInfo
+}
+
+var _ service.TCPConnMetrics = (*tcpConnMetrics)(nil)
+
+func newTCPConnMetrics(tcpCollector *tcpCollector, tunnelTimeCollector *tunnelTimeCollector, clientConn net.Conn, clientInfo ipinfo.IPInfo) *tcpConnMetrics {
+	tcpCollector.openConnection(clientInfo)
+	return &tcpConnMetrics{
+		tcpCollector:        tcpCollector,
+		tunnelTimeCollector: tunnelTimeCollector,
+		localAddr:           clientConn.LocalAddr(),
+		clientAddr:          clientConn.RemoteAddr(),
+		clientInfo:          clientInfo,
+	}
+}
+
+func (cm *tcpConnMetrics) AddAuthenticated(accessKey string) {
+	ipKey, err := toIPKey(cm.clientAddr, accessKey)
+	if err == nil {
+		cm.tunnelTimeCollector.startConnection(*ipKey)
+	}
+}
+
+func (cm *tcpConnMetrics) AddClosed(accessKey string, status string, data metrics.ProxyMetrics, duration time.Duration) {
+	cm.tcpCollector.proxyCollector.addOutbound(data.ClientProxy, data.ProxyTarget, accessKey, cm.clientInfo)
+	cm.tcpCollector.proxyCollector.addInbound(data.TargetProxy, data.ProxyClient, accessKey, cm.clientInfo)
+	cm.tcpCollector.closeConnection(status, duration, accessKey, cm.clientInfo)
+	ipKey, err := toIPKey(cm.clientAddr, accessKey)
+	if err == nil {
+		cm.tunnelTimeCollector.stopConnection(*ipKey)
+	}
+}
+
+func (cm *tcpConnMetrics) AddProbe(status, drainResult string, clientProxyBytes int64) {
+	cm.tcpCollector.addProbe(cm.localAddr.String(), status, drainResult, clientProxyBytes)
+}
+
 type tcpCollector struct {
```

Comment:
Should we call this `tcpServiceMetrics` (or `Endpoint`?) to align with `tcpConnMetrics`? Collector/Metrics is inconsistent. I think Metrics is a bit more intuitive.